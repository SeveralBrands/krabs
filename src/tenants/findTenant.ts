import { Tenant, DomainSpec } from '../config/config.d';

export function findTenant(tenants: Tenant[], hostname: string): Tenant | undefined {
  return tenants.find((tenant: Tenant) => {
    const domains = tenant.domains.reduce(
      // @ts-ignore
      (acc, current) => [...acc, ...Object.values(current)] as never,
      [],
    ) as DomainSpec[];

    if (domains.includes(hostname)) {
      return true;
    } else {
      const regexDomains = domains.filter(
        (domain: DomainSpec) => domain instanceof RegExp,
      ) as RegExp[];
      const match = regexDomains.map((domain: RegExp) => domain.test(hostname)).filter(Boolean);
      return match.length;
    }
  });
}

export default findTenant;
