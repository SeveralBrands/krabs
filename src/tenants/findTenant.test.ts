import { findTenant } from './findTenant';

describe('findTenant', () => {
  it('should find the right tenant', () => {
    const config = [
      {
        name: 'website-1',
        domains: [
          {
            dev: 'local.website-1.com',
            test: 'test.website-1.com',
          },
          {
            development: 'local.dev.website-2.com',
            prod: 'prod.website-2.com',
            test: 'test.website-2.com',
          },
        ],
      },
      {
        name: 'website-2',
        domains: [
          {
            dev: /local\.website-2\.[a-z]*\.com/,
            test: /test\.website-2\.[a-z]*\.com/,
          },
        ],
      },
    ];

    expect(findTenant(config, 'test.website-1.com')).toMatchSnapshot();
    expect(findTenant(config, 'test.website-2.example.com')).toMatchSnapshot();
    expect(findTenant(config, 'test.website-2.foo.com')).toMatchSnapshot();
    expect(findTenant(config, 'test.website-2.bar.com')).toMatchSnapshot();
    expect(findTenant(config, 'test.website-2.krabs.com')).toMatchSnapshot();
  });
});
