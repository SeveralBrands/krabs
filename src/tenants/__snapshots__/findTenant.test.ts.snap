// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findTenant should find the right tenant 1`] = `
Object {
  "domains": Array [
    Object {
      "dev": "local.website-1.com",
      "test": "test.website-1.com",
    },
    Object {
      "development": "local.dev.website-2.com",
      "prod": "prod.website-2.com",
      "test": "test.website-2.com",
    },
  ],
  "name": "website-1",
}
`;

exports[`findTenant should find the right tenant 2`] = `
Object {
  "domains": Array [
    Object {
      "dev": /local\\\\\\.website-2\\\\\\.\\[a-z\\]\\*\\\\\\.com/,
      "test": /test\\\\\\.website-2\\\\\\.\\[a-z\\]\\*\\\\\\.com/,
    },
  ],
  "name": "website-2",
}
`;

exports[`findTenant should find the right tenant 3`] = `
Object {
  "domains": Array [
    Object {
      "dev": /local\\\\\\.website-2\\\\\\.\\[a-z\\]\\*\\\\\\.com/,
      "test": /test\\\\\\.website-2\\\\\\.\\[a-z\\]\\*\\\\\\.com/,
    },
  ],
  "name": "website-2",
}
`;

exports[`findTenant should find the right tenant 4`] = `
Object {
  "domains": Array [
    Object {
      "dev": /local\\\\\\.website-2\\\\\\.\\[a-z\\]\\*\\\\\\.com/,
      "test": /test\\\\\\.website-2\\\\\\.\\[a-z\\]\\*\\\\\\.com/,
    },
  ],
  "name": "website-2",
}
`;

exports[`findTenant should find the right tenant 5`] = `
Object {
  "domains": Array [
    Object {
      "dev": /local\\\\\\.website-2\\\\\\.\\[a-z\\]\\*\\\\\\.com/,
      "test": /test\\\\\\.website-2\\\\\\.\\[a-z\\]\\*\\\\\\.com/,
    },
  ],
  "name": "website-2",
}
`;
